Source/WebCore/platform/graphics/ImageSource.cpp

```diff
@@ -26,705 +26,23 @@
-
-    if (auto expectedContentSize = expectedContentLength())
-        m_decoder->setExpectedContentSize(expectedContentSize);
-
-    // Changing the decoder has to stop the decoding thread. The current frame will
-    // continue decoding safely because the decoding thread has its own
-    // reference of the old decoder.
-    stopAsyncDecodingQueue();
-    return true;
-}
-
-void ImageSource::setData(FragmentedSharedBuffer* data, bool allDataReceived)
-{
-    if (!data || !ensureDecoderAvailable(data))
-        return;
-
-    m_decoder->setData(*data, allDataReceived);
-}
-
-void ImageSource::resetData(FragmentedSharedBuffer* data)
-{
-    m_decoder = nullptr;
-    setData(data, isAllDataReceived());
-}
-
-EncodedDataStatus ImageSource::dataChanged(FragmentedSharedBuffer* data, bool allDataReceived)
-{
-    setData(data, allDataReceived);
-    clearMetadata();
-    EncodedDataStatus status = encodedDataStatus();
-    if (status >= EncodedDataStatus::SizeAvailable)
-        growFrames();
-    return status;
-}
-
-bool ImageSource::isAllDataReceived()
-{
-    return isDecoderAvailable() ? m_decoder->isAllDataReceived() : frameCount();
-}
-
-void ImageSource::destroyDecodedData(size_t begin, size_t end)
-{
-    if (begin >= end)
-        return;
-
-    ASSERT(end <= m_frames.size());
-
-    unsigned decodedSize = 0;
-
-    for (size_t index = begin; index < end; ++index)
-        decodedSize += m_frames[index].clearImage();
-
-    decodedSizeReset(decodedSize);
-}
-
-void ImageSource::destroyIncompleteDecodedData()
-{
-    unsigned decodedSize = 0;
-
-    for (auto& frame : m_frames) {
-        if (!frame.hasMetadata() || frame.isComplete())
-            continue;
-
-        decodedSize += frame.clear();
-    }
-
-    decodedSizeDecreased(decodedSize);
-}
-
-void ImageSource::clearFrameBufferCache(size_t beforeFrame)
-{
-    if (!isDecoderAvailable())
-        return;
-    m_decoder->clearFrameBufferCache(beforeFrame);
-}
-
-void ImageSource::encodedDataStatusChanged(EncodedDataStatus status)
-{
-    if (status == m_encodedDataStatus)
-        return;
-
-    m_encodedDataStatus = status;
-
-    if (status >= EncodedDataStatus::SizeAvailable)
-        growFrames();
-
-    if (!m_image)
-        return;
-
-    if (auto observer = m_image->imageObserver())
-        observer->encodedDataStatusChanged(*m_image, status);
-}
-
-void ImageSource::decodedSizeChanged(long long decodedSize)
-{
-    if (!decodedSize || !m_image)
-        return;
-
-    if (auto imageObserver = m_image->imageObserver())
-        imageObserver->decodedSizeChanged(*m_image, decodedSize);
-}
-
-void ImageSource::decodedSizeIncreased(unsigned decodedSize)
-{
-    if (!decodedSize)
-        return;
-
-    m_decodedSize += decodedSize;
-
-    // The fully-decoded frame will subsume the partially decoded data used
-    // to determine image properties.
-    long long changeSize = static_cast<long long>(decodedSize) - m_decodedPropertiesSize;
-    m_decodedPropertiesSize = 0;
```

Comment:
I don't see this logic to subtract the size for the decoded properties in the equivalent function in the new code (ImageFrameProvider::decodedSizeIncreased). Is that a deliberate change? It looks like ImageSource will think the final image size will be the decoded properties size plus the decoded image size. Maybe that's not a big deal, since the decoded properties size is never going to be that big?

Reply:
I put this code back in `ImageFrameProvider::decodedSizeIncreased()`

Reply:
I think there is a problem with the order we do this, which I'll comment on in the new patch.